rest-server/src/config/launcher.js

```diff
@@ -92,6 +92,7 @@ if (error) {
 }
 launcherConfig = value;
 
+// TODO: Replace hdfs commands with webhdfs APIs?
```

Comment:
why not use native hdfs?

Reply:
in the future, to improve modularity, we need to make this file system agnostic. so users can replace hdfs with whatever storage system they want. 

---
In reply to: [158621888](https://github.com/Microsoft/pai/pull/134#discussion_r158621888) [](ancestors = 158621888)

Reply:
changing from native hdfs to web hdfs api doesn't achieve this

Reply:
Rest-server is rest-server. Using the native hdfs binary introduces too many dependencies between modules -- you have to prepare the hdfs binary in the image at least. In addition, we just want to create directories here, it's not efficient to use such a heavy thing.

Reply:
rest-server is part of the infrastructure and it is inside the hadoop cluster. webhdfs uses http and needs to access all nodes which will limit the performance, so why should we use the http api instead of the shell command? the cluster prepared hadoop binary during deployment and other pods like frameworklauncher also needs it.

Reply:
hdfs does not need to access data node? what is its mechanism? in addition, in micro-service architecture, http apis are the recommended way to interact with other micro service components (see: https://en.m.wikipedia.org/wiki/Microservices). correct me if my understanding is wrong.